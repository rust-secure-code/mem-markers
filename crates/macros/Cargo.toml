[package]
name = "macros"
version = "0.1.0"
authors = ["Ryan Levick <ryan.levick@gmail.com>"]
edition = "2018"

[lib]
proc_macro = true

[[test]]
name = "tests"
path = "tests/diagnostics.rs"

[dependencies]
quote = "1.0"
syn = "1.0"
proc-macro2 = "1.0"

[dev-dependencies]
trybuild = "1.0"
mem-markers = { path = "../.." }
