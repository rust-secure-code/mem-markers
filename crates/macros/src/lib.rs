extern crate proc_macro;
use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, DeriveInput};

/// Derive a Fixed layout for a type
///
/// ```
/// #[repr(C)]
/// struct Foo {
///     bar: u8
/// }
/// ```
#[proc_macro_derive(FixedLayout)]
pub fn derive(item: TokenStream) -> TokenStream {
    let input = parse_macro_input!(item as DeriveInput);

    let type_name = input.ident;
    let can_derive = input.attrs.iter().any(|attr| {
        let meta = attr.parse_meta().unwrap();
        match meta {
            syn::Meta::List(list) => {
                if !list.path.is_ident("repr") {
                    return false;
                }
                let first = list.nested.first().unwrap();
                match first {
                    syn::NestedMeta::Meta(syn::Meta::Path(p)) if p.is_ident("C") => {
                        return true;
                    }
                    _ => return false,
                }
            }
            _ => false,
        }
    });

    if !can_derive {
        panic!("NO!")
    }

    (quote! {
        unsafe impl mem_markers::FixedLayout for #type_name {}
    })
    .into()
}
