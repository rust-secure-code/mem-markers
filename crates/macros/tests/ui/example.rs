use macros::FixedLayout;
// use mem_markers::FixedLayout;

#[derive(FixedLayout)]
struct Foo {
    f: u8,
}

#[derive(FixedLayout)]
#[repr(C)]
struct Bar {
    b: u8,
}

fn main() {}
