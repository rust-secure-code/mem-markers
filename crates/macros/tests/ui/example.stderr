error: proc-macro derive panicked
 --> $DIR/example.rs:4:10
  |
4 | #[derive(FixedLayout)]
  |          ^^^^^^^^^^^
  |
  = help: message: NO!
